testing/multi-env-test-engine/src/main/java/org/projectnessie/junit/engine/MultiEnvTestDescriptorTree.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright (C) 2023 Dremio
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.junit.engine;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import org.junit.jupiter.engine.config.CachingJupiterConfiguration;
+import org.junit.jupiter.engine.config.JupiterConfiguration;
+import org.junit.jupiter.engine.descriptor.JupiterEngineDescriptor;
+import org.junit.jupiter.engine.discovery.DiscoverySelectorResolver;
+import org.junit.platform.engine.ConfigurationParameters;
+import org.junit.platform.engine.EngineDiscoveryRequest;
+import org.junit.platform.engine.TestDescriptor;
+import org.junit.platform.engine.UniqueId;
+import org.junit.platform.engine.UniqueId.Segment;
+
+public class MultiEnvTestDescriptorTree {
+
+  private static final boolean FAIL_ON_MISSING_ENVIRONMENTS =
+      !Boolean.getBoolean("org.projectnessie.junit.engine.ignore-empty-environments");
+
+  private final TestDescriptor rootNode;
+  private final ConfigurationParameters configurationParameters;
+  private final AtomicBoolean foundAtLeastOneEnvironmentId = new AtomicBoolean();
+  private final List<String> processedExtensionNames = new ArrayList<>();
+  private List<TestDescriptor> latestLeafNodes;
+
+  public MultiEnvTestDescriptorTree(
+      TestDescriptor rootNode, ConfigurationParameters configurationParameters) {
+    this.rootNode = rootNode;
+    this.latestLeafNodes = Collections.singletonList(rootNode);
+    this.configurationParameters = configurationParameters;
+  }
+
+  public void appendDescriptorsForExtension(MultiEnvTestExtension multiEnvTestExtension) {
+    processedExtensionNames.add(multiEnvTestExtension.getClass().getSimpleName());
+
+    List<TestDescriptor> newLeafNodes = new ArrayList<>();
+    for (TestDescriptor parentNode : latestLeafNodes) {
+      for (String environmentId :
+          multiEnvTestExtension.allEnvironmentIds(configurationParameters)) {
```

Comment:
TBH, I find this method of constructing a Cartesian product of environments a bit obscure. Why not simply have nested loops in one method (without calling out to this helper class)?